Tests/Simulation/SimulatedLoadTests.cs

```diff
@@ -119,10 +119,121 @@ async Task ActAsync(SimulatedDevice device, TimeSpan startOffset)
+            var simulatedOtaaDevices = InitializeSimulatedDevices(TestFixtureSim.DeviceRange3000_OTAA_FullLoad);
+            Assert.Equal(simulatedAbpDevices.Count, simulatedOtaaDevices.Count);
+            Assert.True(simulatedOtaaDevices.Count < 50, "Simulator does not work for more than 50 of each devices (due to IoT Edge connection mode). To go beyond 100 device clients, use edge hub environment variable 'MaxConnectedClients'.");
+            Assert.True(messagesBeforeConfirmed <= messagesBeforeJoin, "OTAA devices should send all messages as confirmed messages.");
+
+            // 1. picking devices send an initial message (warm device cache in LNS module)
+            foreach (var devices in simulatedAbpDevices.Batch(warmUpDeviceStepSize))
+            {
+                await Task.WhenAll(from device in devices select SendWarmupMessageAsync(device));
+                await Task.Delay(delayWarmup);
+            }
+
+            async Task SendWarmupMessageAsync(SimulatedDevice device)
+            {
+                await Task.Delay(TimeSpan.FromMilliseconds(RandomNumberGenerator.GetInt32(10, 250)));
+                this.logger.LogInformation("[WARM-UP] {DeviceId}", device.LoRaDevice.DeviceID);
+                // warm-up messages should not contain the message fragment (we do not want to include them in the message count)
+                using var request = WaitableLoRaRequest.CreateWaitableRequest(device.CreateConfirmedDataUpMessage("foo"));
+                await device.SendDataMessageAsync(request);
+            }
+
+            // 2. ABP and OTAA devices send messages
+            const int initialMessageId = 1;
+            for (var messageId = initialMessageId; messageId < initialMessageId + scenarioMessagesPerDevice; ++messageId)
+            {
+                foreach (var batch in simulatedAbpDevices.Zip(simulatedOtaaDevices, (first, second) => (Abp: first, Otaa: second))
+                                                         .Batch(scenarioDeviceStepSize))
+                {
+                    var payload = this.uniqueMessageFragment + messageId.ToString(CultureInfo.InvariantCulture).PadLeft(3, '0');
+                    var abpTasks = batch.Select(devices => SendUpstreamMessage(devices.Abp, payload, messageId));
+                    var otaaTasks = messageId switch
+                    {
+                        messagesBeforeJoin + initialMessageId => batch.Select(devices => JoinAsync(devices.Otaa)),
+                        > messagesBeforeJoin + initialMessageId => batch.Select(devices => SendUpstreamMessage(devices.Otaa, payload, messageId)),
+                        _ => Array.Empty<Task>()
+                    };
+
+                    await Task.WhenAll(abpTasks.Concat(otaaTasks));
+                    await Task.Delay(IntervalBetweenMessages);
+                }
+
+                static async Task SendUpstreamMessage(SimulatedDevice device, string payload, int messageId)
```

Comment:
Not sure about this but method declarations should be cached, right? Asking because this is inside a for-loop.

Reply:
Checked this in Sharplab, it is indeed converted to just a class method. I still moved the static local functions out of the loop, it is less distracting indeed.